crates/runtime-datafusion-udfs/src/embed.rs

```diff
@@ -0,0 +1,414 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//! [`ScalarUDFImpl`] definitions for embedding function.
+
+use arrow::array::{ListBuilder, PrimitiveBuilder};
+use arrow::datatypes::Float32Type;
+use arrow_schema::{DataType, Field};
+use async_openai::types::EmbeddingInput;
+use datafusion::common::cast::as_string_array;
+use datafusion::error::DataFusionError;
+use datafusion::logical_expr::{DocSection, Documentation, ScalarFunctionArgs};
+use datafusion::scalar::ScalarValue;
+use datafusion::{
+    common::{Result as DataFusionResult, exec_err},
+    logical_expr::{ColumnarValue, ScalarUDFImpl, Signature, TypeSignature, Volatility},
+};
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::{Arc, LazyLock};
+use tokio::sync::RwLock;
+
+pub static EMBED_UDF_NAME: &str = "embed";
+pub static DOCUMENTATION: LazyLock<Documentation> = LazyLock::new(|| Documentation {
+    doc_section: DocSection::default(),
+    description: "Generates embeddings for text using a specified embedding model".to_string(),
+    syntax_example: "embed(text, model_name)".to_string(),
+    sql_example: Some("SELECT embed('hello world', 'potion_2m')".to_string()),
+    arguments: Some(vec![
+        ("text".to_string(), "The text string to embed.".to_string()),
+        (
+            "model_name".to_string(),
+            "The name of the embedding model to use as defined in the Spicepod.".to_string(),
+        ),
+    ]),
+    alternative_syntax: Some(vec!["embed(['foo', 'bar'], 'potion_2m')".to_string()]),
+    related_udfs: None,
+});
+
+pub static SIGNATURE: LazyLock<Signature> = LazyLock::new(|| {
+    Signature::one_of(
+        // In order of least likely to auto-coerce, via logical_expr docs for OneOf:
+        // > Coercion is attempted to match the signatures in order,
+        // > and stops after the first success, if any.
+        vec![
+            // embed(make_array(a, b, c), model_name)
+            TypeSignature::Exact(vec![
+                DataType::List(Arc::new(Field::new("sentence", DataType::Utf8, true))),
```

Comment:
This will type coerce for `DataType::FixedSizedList(LargeUTF8...`?

Reply:
Thanks for pointing this out. Looks like it won't: I added some tests and tried to support every combination of list/string.

Re this note: apparently if you invert the order of the type signatures the UDF breaks because DF will apparently cast your list or array into a `"[string, literal]" 🤷‍♀️